### ISSUE TITLE
GLU backend test contains Split with no split attr

### ISSUE THREAD

**Issue by tjingrant** [onnx/onnx#349]
Quote from the Spec for your reference:

> Split a tensor into a list of tensors, along the specified 'axis'. The lengths of the split can be specified using argument 'axis' or optional second input blob to the operator. Otherwise, the tensor is split to equal sized parts.

I have two questions, what kind of situation does `otherwise` refer to? Is it when `split` is absent? Then how many equally sized parts? I'm asking because I saw this split op in GLU test:
```
input: "0"
output: "1"
output: "2"
op_type: "Split"
attribute {
  name: "axis"
  i: -1
  type: INT
}
```
It is not at all clear what this op is trying to do (if you don't look at the number of outputs...).
cc @bddppq @dzhulgakov

**Comment by bddppq** [comment]
It will then based on the number of outputs to equally split the input tensor.

**Comment by tjingrant** [comment]
@bddppq this would work for now, but looks like a convoluted way to get the split size. I believe in Torch and Caffe2 (and TF as well), an equivalent of `split` has to be supplied to the op. The natural place for this `split` parameter to be is the attributes of the node proto. It seems like now we are just recovering this piece of information lost in the translation. We might as well keep it in the first place.

**Comment by bddppq** [comment]
This is used to express "always split my input into n parts", where `n` is fixed but the number of elements in each part is dynamic, thus is unable to use `split` attribute to express.
Caffe2 also supports this. TF use a slightly different way in that `split` can be either a list of ints or a single int (which will be treated as the number of outputs).

**Comment by tjingrant** [comment]
I understand what you are trying to achieve here and agree that it can be done easily with TF. What bothers me here is that the computation of an operation to perform relies on information about its outputs. This is like overloading by return types in C/C++ and should be avoided.

That is to say, just the number of outputs specified alone can actually change the behavior of the algorithm. This seems like a bad idea. Output shapes/types can be used to ensure correctness (like type checking), but should not be used to determine computation. This is conceptually true for most programming languages like C/C++ in which return type information of a function alone cannot determine/change the behavior of the function. Even if that is the default behavior of PyTorch/Caffe, I believe we should make it right here. 

Specifically, Tensorflow split makes a bit more sense (although arguably hacky nontheless):
https://www.tensorflow.org/api_docs/python/tf/split. We can pass that `n` you mentioned as the `num_or_size_splits` argument. I'd propose to modify the Spec for split to make room for another argument (or extend the purpose of the split argument to be like TF).

**Comment by ezyang** [comment]
> This is like overloading by return types in C/C++ (which is an impossibility) and should be avoided.

ONNX is not C++, and thus different design constraints apply. In the case of ONNX, every operator always has an explicit list of outputs, so there is no problem with overloads. The analogous situation in C++ would be if you never "returned" from a function, but instead passed in references for outputs (e.g., `f(int in1, int in2, int& out1, int& out2)`); in this case, it is clear that overloading can be made to work. The TF split situation is different, because here split is a Python function to generate a node in question.

Now, the internal representation doesn't necessarily correspond one-to-one with how you should implement a DSL for defining ONNX nodes. In the case of PyTorch, we have an `outputs` parameter which specifies how many outputs you want to create an ONNX node with.

**Comment by tjingrant** [comment]
@ezyang the nature of the discussion I posted was about naturalness. On one hand, your example is still overloading by inputs (albeit they are pointers to the outputs) in its simplest level of abstraction (i.e., a parser's prospective, thus the most conducive to interpretation); on the other, I was unable to see how your example workaround to achieve "overloading by return types" address my concern.. That is to say, possibility != naturalness. 

> Now, the internal representation doesn't necessarily correspond one-to-one with how you should implement a DSL for defining ONNX nodes. 

This is exactly my question. Why bother having different representations? I can only see that your point about PyTorch adds to the up side of having a unified way of defining ONNX nodes and generating ONNX nodes in the exporter/importer. What if the user wants to edit the graph exported from PyTorch and could not find the equivalent of `outputs` parameter in the ONNX graph proto?
